src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Pal/Unix/DecryptorPalOpenSsl.Decrypt.cs

```diff
@@ -0,0 +1,123 @@
+    {
+        public sealed override ContentInfo TryDecrypt(
+            RecipientInfo recipientInfo,
+            X509Certificate2 cert,
+            X509Certificate2Collection originatorCerts,
+            X509Certificate2Collection extraStore,
+            out Exception exception)
+        {
+            Debug.Assert(recipientInfo != null);
+            Debug.Assert(cert != null);
+            Debug.Assert(originatorCerts != null);
+            Debug.Assert(extraStore != null);
+
+            string recipiendKekAlgo = recipientInfo.KeyEncryptionAlgorithm.Oid.Value;
+            string certOid = cert.GetKeyAlgorithm();
+
+            ContentInfo decryptedContent = null;
+            RecipientInfoType type = recipientInfo.Type;
+
+            switch (type)
+            {
+                case RecipientInfoType.KeyTransport:
+                    decryptedContent = TryDecryptTrans(cert, out exception);
+                    break;
+                        
+                case RecipientInfoType.KeyAgreement:
+                    // OpenSSL doesn't seem to support KeyAgreement directly, the only way to work around this seems to be to extract
+                    // the information by parsing it and to store the appropriate keys to call the EVP_PKEY_derive method on them
+                    // to obtain the secret and proceed with manual decryption. For further documentation see:
+                    // https://wiki.openssl.org/index.php/EVP_Key_Agreement
+                    exception = new PlatformNotSupportedException();
+                    break;
+
+                default:
+                    // As we commented in the Windows implementation:
+                    // Since only the framework can construct RecipientInfo's, we're at fault if we get here. So it's okay to assert and throw rather than
+                    // returning to the caller.
+                    Debug.Fail($"Unexpected RecipientInfoType: {type}");
+                    throw new NotSupportedException();
+            }
+
+            if (exception != null)
+                return null;
+
+            return decryptedContent;
+        }
+
+        private ContentInfo TryDecryptTrans(X509Certificate2 recipientCert, out Exception exception)
+        {
+            // If there's no content OpenSSL will fail to decrypt, so do the check manually before
+            // delegating to OpenSSL
+            ContentInfo currentContent = _decodedMessage.GetEmbeddedContent();
+            if (currentContent.Content.Length == 0)
+            {
+                exception = null;
+                return currentContent;
+            }
+
+            byte[] decryptedMessage = null;
+
+            using (RSAOpenSsl rsa = recipientCert.GetRSAPrivateKey() as RSAOpenSsl)
+            using (SafeBioHandle decryptedBuffered = Interop.Crypto.CreateMemoryBio())
+            using (SafeX509Handle recipientCertHandle = Interop.Crypto.X509Duplicate(recipientCert.Handle))
```

Comment:
Naming: decryptedBuffered is probably more accurately decryptedData, or decryptionBuffer

And these two SafeHandles can be moved down to the inner using, since we shouldn't create/destroy them if the RSA object was unrecognized.

And the using on the RSA object should really be

```C#
using (RSA rsa = recipientCert.GetRSAPrivateKey())
{
    RSAOpenSsl rsaOpenSsl = rsa as RSAOpenSsl;

    if (rsaOpenSsl == null)
    {
        ...
    }

    ...
```

So that if it is not-null-and-not-RSAOpenSsl it still gets Disposed.